src/main/kotlin/snyk/common/lsp/LanguageServerWrapper.kt

```diff
@@ -133,8 +134,36 @@ class LanguageServerWrapper(
         return workspaceFolders.toList()
     }
 
-    fun getWorkspaceFolders(project: Project) =
-        project.getContentRootVirtualFiles().mapNotNull { WorkspaceFolder(it.url, it.name) }.toSet()
+    fun getWorkspaceFolders(project: Project): Set<WorkspaceFolder> {
+        val normalizedRoots = getTrustedContentRoots(project)
+        return normalizedRoots.map { WorkspaceFolder(it.url, it.name) }.toSet()
+    }
+
+    private fun getTrustedContentRoots(project: Project): MutableSet<VirtualFile> {
+        // the sort is to ensure that parent folders come first
+        // e.g. /a/b should come before /a/b/c
+        val contentRoots = project.getContentRootVirtualFiles().filterNotNull().sortedBy { it.path }
+        val trustService = service<WorkspaceTrustService>()
+        val normalizedRoots = mutableSetOf<VirtualFile>()
+
+        for (root in contentRoots) {
+            val pathTrusted = trustService.isPathTrusted(root.toNioPath())
+            if (!pathTrusted) {
+                logger.debug("Path not trusted: ${root.path}")
+                continue
+            }
+
+            var add = true
+            for (normalizedRoot in normalizedRoots) {
+                if (!root.path.startsWith(normalizedRoot.path)) continue
+                add = false
+                break
+            }
+            if (add) normalizedRoots.add(root)
+        }
+        logger.debug("Normalized content roots: $normalizedRoots")
+        return normalizedRoots
+    }
```

Comment:
Does this work fine with super nested folders? Is there a way to check it?

Reply:
It does, the testing involves a lot of mocking of IntelliJ APIs though. It came to a point, that I deleted the test, because I had to know the implementation in detail. Maybe writing a LightPlatformTest would solve that instead of mocking. I'll look into it.